packages/react-reconciler/src/ReactFiberBeginWork.js

```diff
@@ -1705,6 +1705,21 @@ function updateHostHoistable(
         workInProgress,
       );
     }
+  } else {
+    // Once a HostHoistable is a certain "type" like an Instance vs a Resource
+    // we can't change it's type. For now we error because this is an edge case and
+    // should happen rarely. In the future we will implement an inner type so we
+    // can change from one implementation to the next during an update
+    if (
+      // This Fiber is trying to go from Instance to Resource
+      (resource && current.memoizedState === null) ||
+      // This Fiber is trying to go from Resource to Instance
+      (resource === null && current.memoizedState)
+    ) {
+      throw new Error(
+        'Expected HostHoistable to not change between Instance and Resource type. Try adding a key to this component so that when its props change a new instance is mounted.',
```

Comment:
This is a public error message likely to be hit by an end user yet it includes three Fiber internal terminologies.

Reply:
We should still try to express it in terms of the possible mistakes that could lead it to happen. It can link to a docs page if the explanation is too verbose.

Reply:
Re: host-specific language, it's annoying but the right thing to do is probably a fiber config method.